salt/states/file.py

```diff
@@ -352,6 +352,33 @@ def _check_user(user, group):
     return err
 
 
+def _is_valid_relpath(
+        relpath,
+        maxdepth=None):
+    '''
+    Performs basic sanity checks on a relative path.
+
+    Requires slash-separated path components (i.e. the kind obtained
+    through cp.list_master or other such calls).
+
+    Ensures that the path does not contain directory transversal, and
+    that it does not exceed a stated maximum depth (if specified).
+    '''
+    # Check relpath surrounded by slashes, so that `..` can be caught as
+    # a path component at the start, end, and in the middle of the path.
+    if '/../' in '/{0}/'.format(relpath):
```

Comment:
The slash after (and before) the `..` is to ensure that it is a whole path component. So that this picks up `../some_path` too (and other cases), the whole path is first surrounded in slashes. This is easier than a test for `(^|/)..($|/)` but otherwise the same.

The original test used `startswith("..")` which has the side effect of rejecting too many paths (any files that start with `..`).

Regarding Windows compatibility, I'm not sure how to proceed. I don't have any Windows machine to test on (and python doesn't seem to work on Wine). All I know is that Windows accepts both `/` and `\` as a path separator, meaning that `os.sep` must be incorrect (because it has to pick only one), and I don't know whether or not `os.path.relpath` normalises all path separators.

For clarity, my main worry is because these paths are identical on Windows:

```
../path/file
..\path\file
..\path/file
../path\file
```

Approaches and their problems:

* Deal with `os.sep`
  * Forward-slashes break despite being valid
  * Does relpath normalise to one or the other?
* Split parts of the function into "windows" and "everyone else" versions
  * "everyone else" being the `os.sep` version
* Replace backslashes with forward slashes
  * Backslashes are a valid filename character on other operating systems.
  * Is there an OS which uses backslashes which doesn't support forward slashes?
  * Probably a bad idea all around
* Write a different `relpath` that works more sanely (i.e. deals with forward-slash-separated paths only) and transform into an `os.sep`-separated path at the 
  * Might not be necessary depending on how Windows' `relpath` behaves
* Use `os.sep` anyway and let windows users complain when it doesn't work
  * I like this idea :^)

I'm not sure which option to take here.
It's also worth pointing out that the original code made the same assumptions about platform and path format.